### PR TITLE
Tx Pool tests + parameter checking

### PR DESCRIPTION
**Pull request by ChristianBorst**
Adding more tests related to the new tx pool indexing model, as well as performing some edge case analysis for the main public methods.

A few public methods would cause nil pointers rather than return errors, so added some checks to handle invalid arguments

### CODE REVIEW

File: module/x/gravity/keeper/pool.go [hunk 1]
```diff
@@ -16,8 +16,20 @@ import (
...
-func (k Keeper) AddToOutgoingPool(ctx sdk.Context, sender sdk.AccAddress, counterpartReceiver string, amount sdk.Coin, fee sdk.Coin) (uint64, error) {
+// - adds the TX to the `available` TX pool
+func (k Keeper) AddToOutgoingPool(
+	ctx sdk.Context,
+	sender sdk.AccAddress,
+	counterpartReceiver string,
+	amount sdk.Coin,
+	fee sdk.Coin,
+) (uint64, error) {
+	if ctx.IsZero() || sender.Empty() || counterpartReceiver == "" ||
```
Comment (ChristianBorst) [inline module/x/gravity/keeper/pool.go:27]: If there is a more elegant way to handle parameter checking in Go I would love to adjust this. Additionally, with this approach we do not tell the method caller which argument was a problem. I have left the parameter checking in this style because it causes less method bloat than individual checks like `if amount.Amount.IsNil() || ... { return 0, "Bad amount argument" }`

File: module/x/gravity/keeper/pool.go [hunk 2]
```diff
@@ -115,6 +130,11 @@ func (k Keeper) RemoveFromOutgoingPoolAndRefund(ctx sdk.Context, txId uint64, se
 	if err != nil {
 		return sdkerrors.Wrapf(types.ErrInvalid, "txId %d not in unbatched index! Must be in a batch!", txId)
 	}
+	// Make sure the tx was removed
```
Comment (ChristianBorst) [inline module/x/gravity/keeper/pool.go:133]: This check may be completely unnecessary, but it should be O(1) and only happens on unbatched transaction cancellation

File: module/x/gravity/keeper/pool_test.go [hunk 3]
```diff
@@ -79,6 +81,70 @@ func TestAddToOutgoingPool(t *testing.T) {
 	assert.Equal(t, exp, got)
 }
 
+// Checks some common edge cases like invalid inputs, user doesn't have enough tokens, token doesn't exist, inconsistent entry
```
Comment (ChristianBorst) [inline module/x/gravity/keeper/pool_test.go:84]: Wall of tests incoming

File: module/x/gravity/keeper/batch.go [hunk 4]
```diff
@@ -162,6 +162,10 @@ func (k Keeper) pickUnbatchedTX(
 		if tx != nil && tx.Erc20Fee != nil {
 			selectedTx = append(selectedTx, tx)
 			err = k.removeUnbatchedTX(ctx, *tx.Erc20Fee, tx.Id)
+			oldTx, oldTxErr := k.GetUnbatchedTxByFeeAndId(ctx, *tx.Erc20Fee, tx.Id)
+			if oldTx != nil || oldTxErr == nil {
+				panic("picked a duplicate transaction from the pool, duplicates should never exist!")
```
Comment (jkilpatr) [inline module/x/gravity/keeper/batch.go:167]: very much like checks like this.

File: module/x/gravity/keeper/pool.go (hunk 1 above)
Comment (jkilpatr) [reply module/x/gravity/keeper/pool.go:27]: the only better way I can think of is having an error checking function specifically for this, just moving the bloat elsewhere where it won't be distracting.

It does seem like the most 'Go' way to do this.

Comment (jkilpatr) [reply module/x/gravity/keeper/pool.go:27]: For several of these (for example the coin type) there could be verifier functions that we get to reuse here.

File: module/x/gravity/keeper/pool.go [hunk 5]
```diff
@@ -265,7 +287,7 @@ func (k Keeper) GetBatchFeeByTokenType(ctx sdk.Context, tokenContractAddr string
 		}
 		batchFee.TotalFees = batchFee.TotalFees.Add(fee.Amount)
 		txCount += 1
-		return txCount < int(maxElements)
+		return txCount == int(maxElements)
```
Comment (jkilpatr) [inline module/x/gravity/keeper/pool.go:290]: if I'm interpreting this right the closure exits when the function we pass it returns false, so == will return false immediately and not give us the correct amount in fees.

File: module/x/gravity/keeper/pool_test.go [hunk 6]
```diff
@@ -142,3 +208,482 @@ func TestTotalBatchFeeInPool(t *testing.T) {
...
+
+	// Check that removing a transaction refunds the costs and the tx no longer exists in the pool
+	checkRemovedTx(t, input, ctx, ids[2], fees[2], amounts[2], &feesAndAmounts, originalBal, mySender, myTokenContractAddr, myTokenDenom)
+	checkRemovedTx(t, input, ctx, ids[3], fees[3], amounts[3], &feesAndAmounts, originalBal, mySender, myTokenContractAddr, myTokenDenom)
+	checkRemovedTx(t, input, ctx, ids[1], fees[1], amounts[1], &feesAndAmounts, originalBal, mySender, myTokenContractAddr, myTokenDenom)
+	checkRemovedTx(t, input, ctx, ids[0], fees[0], amounts[0], &feesAndAmounts, originalBal, mySender, myTokenContractAddr, myTokenDenom)
+	require.Empty(t, input.GravityKeeper.GetUnbatchedTransactions(ctx))
+}
+
+func checkRemovedTx(t *testing.T, input TestInput, ctx sdk.Context, id uint64, fee uint64, amount uint64,
```
Comment (jkilpatr) [inline module/x/gravity/keeper/pool_test.go:349]: this function needs a little more commenting.

**Review by jkilpatr — CHANGES_REQUESTED** [review]
This looks great! really happy with the changes here and the extra checks added.

I've got two areas of concern where I've left comments.

File: module/x/gravity/keeper/pool.go (hunk 5 above)
Comment (ChristianBorst) [reply module/x/gravity/keeper/pool.go:290]: It `break`s when we return true:

```
if cb(iter.Key(), &transact) {
	break
}
```
We want to return false until we're done iterating

Comment (jkilpatr) [reply module/x/gravity/keeper/pool.go:290]: ah in that case I was confused when I wrote this and it's a good correction.

Thanks!

**Review by jkilpatr — APPROVED** [review]
